harness/iteratorZipUtils.js

```diff
@@ -0,0 +1,223 @@
+
+  var ownKeys = Reflect.ownKeys(result);
+  assert.compareArray(ownKeys, ["value", "done"], label + ": iterator result properties");
+
+  verifyProperty(result, "value", {
+    value: value,
+    writable: true,
+    enumerable: true,
+    configurable: true,
+  });
+
+  verifyProperty(result, "done", {
+    value: done,
+    writable: true,
+    enumerable: true,
+    configurable: true,
+  });
+}
+
+// Assert |array| is a packed array with default property attributes.
+function assertIsPackedArray(array, label) {
+  assert(Array.isArray(array), label + ": array is an array exotic object");
+
+  assert.sameValue(
+    Object.getPrototypeOf(array),
+    Array.prototype,
+    label + ": [[Prototype]] of array is Array.prototype"
+  );
+
+  assert(Object.isExtensible(array), label + ": array is extensible");
+
+  // Ensure "length" property has its default property attributes.
+  verifyProperty(array, "length", {
+    writable: true,
+    enumerable: false,
+    configurable: false,
+  });
+
+  // Ensure no holes and all elements have the default property attributes.
+  for (var i = 0; i < array.length; i++) {
+    verifyProperty(array, i, {
+      writable: true,
+      enumerable: true,
+      configurable: true,
+    });
+  }
+}
+
+// Assert |array| is an extensible null-prototype object with default property attributes.
+function _assertIsNullProtoMutableObject(object, label) {
+  assert.sameValue(
+    Object.getPrototypeOf(object),
+    null,
+    label + ": [[Prototype]] of object is null"
+  );
+
+  assert(Object.isExtensible(object), label + ": object is extensible");
+
+  // Ensure all properties have the default property attributes.
+  var keys = Object.getOwnPropertyNames(object);
+  for (var i = 0; i < keys.length; i++) {
+    verifyProperty(object, keys[i], {
+      writable: true,
+      enumerable: true,
+      configurable: true,
+    });
+  }
+}
+
+// Assert that the `zipped` iterator results as for Iterator.zip for the first `count` elements.
+// Assumes inputs is an array of arrays of length >= count.
+// Advances `zipped` by `count` steps.
+function assertZipped(zipped, inputs, count, label) {
+  // Last returned elements array.
+  var last = null;
+
+  for (var i = 0; i < count; i++) {
+    var itemLabel = label + ", step " + i;
+
+    var result = zipped.next();
+    var value = result.value;
+
+    // Test IteratorResult structure.
+    assertIteratorResult(result, value, false, itemLabel);
+
+    // Ensure value is a new array.
+    assert.notSameValue(value, last, itemLabel + ": returns a new array");
+    last = value;
+
+    // Ensure all array elements have the expected value.
+    var expected = inputs.map(function (array) {
+      return array[i];
+    });
+    assert.compareArray(value, expected, itemLabel + ": values");
+
+    // Ensure value is a packed array with default data properties.
+    //
+    // This operation is destructive, so it has to happen last.
```

Comment:
What does it mean for this to be destructive? It doesn't look like it mutates its input.